src/spec/cli.cr

```diff
@@ -0,0 +1,136 @@
+require "./formatter"
+require "option_parser"
+
+module Spec
+  # :nodoc:
+  class CLI
+    # :nodoc:
+    class Options
+      property formatters = Array(Spec::Formatter).new
+      property locations = Array({String, Int32}).new
+
+      property! default_formatter : Spec::Formatter
```

Comment:
I didn't want to introduce default values here, at least not on this refactor. I tried to bring things as close as possible to the original to reduce the PR size and the changes to be reviewed.

Definitely this could be improved moving forward (I already have another draft to improve formatter registration and remove junit-specific code).

Reply:
I got why you did it this way, but then please make the properties nilable because they are.

Reply:
That will force me to change the code in `apply_options` and use local variables instead.

From this:

```crystal
Spec.pattern = options.pattern if options.pattern?
```

To this:

```crystal
if pattern = options.pattern?
  Spec.pattern = pattern
end
```

Perhaps I'm missing something in the approach you're suggesting, so will appreciate if you can provide a code example with the reasoning instead.

Thank you.

Reply:
Well, the reasoning is that if you happend to call `options.default_formatter` (maybe from a tool that interfaces with this spec CLI) you wouldn't be aware that calling this simple accessor might raise a runtime error. This should be avoided. `property!` should only be used for properties that are usually expected to not be nil when the object is used (for example after setting up) and accessing them in the intended way will not raise a nil assertion.

That's not the case here. Each of these options can either have a value of a specific type or `Nil`. That nilable should be expressed in the code and handled appropriately. Your current implementation tries to treat the options as not being nilable.

As for the usage: assigning to a local variable is the proper Crystal way to handle nilable instance variables. I don't see why you would want to avoid that in favour of easily breakable code.

Reply:
Thank you for expressing the reasoning behind your suggestions. Will push changes to correct usage of property.

Cheers.